streams/shared/src/main/scala/scalaz/zio/stream/ZStream.scala

```diff
@@ -379,10 +412,43 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
     override def fold[R2 <: R1, E2 >: E1, B1 >: B, S]: Fold[R2, E2, B1, S] =
       ZManaged.succeedLazy { (s, cont, g) =>
         self.fold[R2, E2, A, S].flatMap(fold => fold(s, cont, (s, a) => f(a).flatMap(g(s, _))))
-
       }
   }
 
+  final def mapMPar[R1 <: R, E1 >: E, B](n: Int)(f: A => ZIO[R1, E1, B]): ZStream[R1, E1, B] =
+    new ZStream[R1, E1, B] {
+      def fold[R2 <: R1, E2 >: E1, B1 >: B, S]: ZStream.Fold[R2, E2, B1, S] =
+        ZManaged.succeedLazy { (s, cont, g) =>
+          for {
+            out <- Queue.bounded[Take[E1, IO[E1, B]]](n).toManaged(_.shutdown)
```

Comment:
This makes me think that we could use a `withBuffer(n: Int)` and `withBufferManaged(n: Int)` that works like `Semaphore#withPermit`.

Reply:
That sounds interesting, how would that work?

Reply:
Something like this:
```scala
object Queue {
  def withBuffer[R, E, A, B](n: Int)(f: Queue[A] => ZIO[R, E, B]): ZIO[R, E, B] =
    Queue.bounded(n) >>= (q => f(q).ensuring(q.shutdown))
}
```
Probably silly.

Reply:
ah not at all, I almost added that, but `Queue.bounded(n).toManaged(_.shutdown)` was fairly short 🧐